ckb-vm-syscall-tracer/src/lib.rs

```diff
@@ -1018,6 +1018,221 @@ enum PartialLocator {
+        }
+
+        impl<$($type: Collector),*> Default for $trace_name<$($type),*> {
+            fn default() -> Self {
+                Self { $($field: None),* }
+            }
+        }
+
+        #[derive(Default, Clone)]
+        pub struct $name<$($type: Collector),*> {
+            $(pub $field: $type),*
+        }
+
+        impl<$($type: Collector),*> $name<$($type),*> {
+            pub fn new($($field: $type),*) -> Self {
+                Self {
+                    $($field),*
+                }
+            }
+        }
+
+        impl<$($type: Collector),*> Collector for $name<$($type),*>
+        {
+            type Trace = $trace_name<$($type),*>;
+
+            fn syscall_generator<DL, M>(
+                vm_id: &VmId,
+                sg_data: &SgData<DL>,
+                vm_context: &VmContext<DL>,
+                data: &Self,
+            ) -> Vec<Box<(dyn Syscalls<M>)>>
+            where
+                DL: CellDataProvider + HeaderProvider + ExtensionProvider + Send + Sync + Clone + 'static,
+                M: SupportMachine + 'static,
+            {
+                let mut r = vec![];
+                $(r.extend($type::syscall_generator(vm_id, sg_data, vm_context, &data.$field));)*
+                r
+            }
+
+            fn preprocess<DL, V, M>(
+                &self,
+                verifier: &TransactionScriptsVerifier<DL, V, M>,
+                script_group: &ScriptGroup,
+                scheduler: &mut Scheduler<DL, V, M>,
+            ) -> Result<(), Error>
+            where
+                DL: CellDataProvider + HeaderProvider + ExtensionProvider + Send + Sync + Clone,
+                V: Clone,
+                M: DefaultMachineRunner,
+            {
+                $(self.$field.preprocess(verifier, script_group, scheduler)?;)*
+                Ok(())
+            }
+
+            fn postprocess<DL, V, M>(&self, scheduler: &mut Scheduler<DL, V, M>) -> Result<(), Error>
+            where
+                DL: CellDataProvider + HeaderProvider + ExtensionProvider + Send + Sync + Clone,
+                V: Clone,
+                M: DefaultMachineRunner,
+            {
+                $(self.$field.postprocess(scheduler)?;)*
+                Ok(())
+            }
+
+            fn seal(self) -> HashMap<CollectorKey, Self::Trace> {
+                let mut r: HashMap<CollectorKey, Self::Trace> = HashMap::new();
+                $(
+                    for (k, v) in self.$field.seal() {
+                        let d = r.entry(k).or_insert_with($trace_name::default);
+                        d.$field = Some(v);
+                    }
+                )*
+                r
+            }
+        }
+    };
+}
+
+define_combine_collector!(CombineCollector2, CombineCollector2Trace, c1: C, c2: D);
+define_combine_collector!(CombineCollector3, CombineCollector3Trace, c1: C, c2: D, c3: E);
+define_combine_collector!(CombineCollector4, CombineCollector4Trace, c1: C, c2: D, c3: E, c4: F);
```

Comment:
Now that you are writing a macro, I would recommend write it from 2 all the way to 16.

Reply:
done